src/public.ts

```diff
@@ -0,0 +1,255 @@
+import { StreamDBConsumerClient } from "./stream-db/streamDbDatabase";
+import { Timestamp } from "./model/timestamp";
+import { execAndReturnWithRetry } from "./utils";
+import axios from "axios";
+import {strict as assert} from "assert";
+import { PausableStream } from "./stream-db/pausableStream";
+
+
+export class StreamClient {
+    private readonly registry: StreamRegistry;
+    private clientCache: Record<string, StreamDBConsumerClient>
+  
+    public constructor(
+      private readonly options: StreamClientOptions,
+    ) {
+      this.registry = options.registry ?? new RemoteStreamRegistry();
+      this.clientCache = {}
+    }
+  
+    public async findOffset(stream: string, params: {height: number, timestamp?: Timestamp}): Promise<Offset | undefined> {
+      // 1. get all endpoints having the stream
+      // 2. try every to find offset
+      // 3. return first found offset
+ 
+      const endpoints = await this.registry.findStreamEndpoints(stream, new Offset("", BigInt(params.height), params.timestamp ?? Timestamp.zero))
+      for (const endpoint of endpoints) {
+        const client = this.getStreamConsumerClient(endpoint.uri)
+        const offset = await client.findOffset(stream, BigInt(params.height || 0), params.timestamp)
```

Comment:
Does `||` work like this? Did you mean to use `??`?

Reply:
Yes, the only difference is that it is slightly redundant it is a logical OR where the value is returned instead of boolean